### PR TITLE
Test Service for latest secure TYPO3 core

### PR DESCRIPTION
**Pull request by jvanderkroon**
Applied patch from https://forge.typo3.org/issues/58828#change-321465… 

#63

### CODE REVIEW

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php [hunk 1]
```diff
@@ -75,10 +77,40 @@ public static function updateLatestTypo3VersionRegistry()
             if (is_array($details) && !empty($details['stable'])) {
                 $stable[$major] = $details['stable'];
             }
+            if (is_array($details) && is_array($details['releases'])) {
+			    $found = false;
```
Comment (IchHabRecht) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: This variable is not necessary. Just set $security[$major] to the latest before the loop.

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php [hunk 2]
```diff
@@ -75,10 +77,40 @@ public static function updateLatestTypo3VersionRegistry()
...
+			    $latestCheckedVersion = '';
```
Comment (IchHabRecht) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: See above. Not necessary.

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php [hunk 3]
```diff
@@ -75,10 +77,40 @@ public static function updateLatestTypo3VersionRegistry()
...
+                    // if major version > latest_lts check if bugfix part of version number == '0';
+                    // in that case this is the security update and stop searching
+                    if (version_compare($major, $latestLts, '>')) {
+                        if ($versionDetails['type'] === 'security') {
+                            $security[$major] = $version;
+                            $found = true;
+                            break;
+                        }
+                        // if versionDetails->type === 'security' then this is the security update; stop searching
+                        if (self::extractBugfixNumberFromVersion($version) === 0) {
```
Comment (IchHabRecht) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: Any reason why you don't use \TYPO3\CMS\Core\Utility\VersionNumberUtility::convertVersionStringToArray?

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php [hunk 4]
```diff
@@ -75,10 +77,40 @@ public static function updateLatestTypo3VersionRegistry()
...
                 $stable[$major] = $details['stable'];
             }
+            if (is_array($details) && is_array($details['releases'])) {
+			    $found = false;
+			    $latestCheckedVersion = '';
+			    foreach ($details['releases'] as $version => $versionDetails) {
+			        $latestCheckedVersion = $version;
+                    // if major version > latest_lts check if bugfix part of version number == '0';
+                    // in that case this is the security update and stop searching
+                    if (version_compare($major, $latestLts, '>')) {
```
Comment (IchHabRecht) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: Isn't the fist version in releases always the one to choose for $major > $lts?

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php [hunk 5]
```diff
@@ -75,10 +77,40 @@ public static function updateLatestTypo3VersionRegistry()
...
+                }
+                if (!$found) {
+			        $security[$major] = $latestCheckedVersion;
+                }
+            }
         }
         \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('TYPO3\CMS\Core\Registry')->set('tx_caretaker', 'TYPO3versions', $max);
         \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('TYPO3\CMS\Core\Registry')->set('tx_caretaker', 'TYPO3versionsStable', $stable);
-
+		\TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('TYPO3\CMS\Core\Registry')->set('tx_caretaker', 'TYPO3versionsSecurity', $security);
```
Comment (IchHabRecht) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: Please split these three lines. First get the registry object, then use it to store all three information.

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php [hunk 6]
```diff
@@ -113,4 +145,19 @@ protected static function curlRequest($requestUrl = false)
...
         return $response;
     }
+
+    /**
+     * extracts the bugfix part of a version number
+     *
+     * @param string $version
+     * @return integer
+     */
+    protected static function extractBugfixNumberFromVersion($version)
```
Comment (IchHabRecht) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: IMHO this is not needed.

**Comment by etobi** [comment]
@jvanderkroon could you please fix the code style issues and check @IchHabRecht 's comments?
I'll gladly review your updated PR.

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 1 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: ✅

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 2 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: ✅

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 3 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: ✅

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 4 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: What is the intention of checking $major > $latestLts at all?

**Comment by etobi** [comment]
patch continued in #88

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 3 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: ✅

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 5 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: ✅

File: Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php (hunk 6 above)
Comment (etobi) [inline Classes/helpers/class.tx_caretaker_LatestVersionsHelper.php]: ✅

**Comment by etobi** [comment]
solves #63

**Pull request closed without merging**